src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -373,5 +381,181 @@ private static void TrimInaccessibleText(ArrayBuilder<SourceText> segments)
                 segments.Add(writer.ToSourceText());
             }
         }
+
+        /// <summary>
+        /// Delegates to SourceTexts within the CompositeText to determine line information.
+        /// </summary>
+        private sealed class CompositeTextLineInfo : TextLineCollection
+        {
+            private readonly CompositeText _compositeText;
+
+            /// <summary>
+            /// The starting line number for the correspondingly indexed SourceTexts in _compositeText.Segments.
+            /// Multiple consecutive entries could indicate the same line number if the corresponding
+            /// segments don't contain newline characters. 
+            /// </summary>
+            /// <remarks>
+            /// This will be of the same length as _compositeText.Segments
+            /// </remarks>
+            private readonly ImmutableArray<int> _segmentLineNumbers;
+
+            // The total number of lines in our _compositeText
+            private readonly int _lineCount;
+
+            public CompositeTextLineInfo(CompositeText compositeText)
+            {
+                var segmentLineNumbers = ArrayBuilder<int>.GetInstance();
```

Comment:
can initialize to capacity right?  or just use an int[] and wrap directly with ImmutableCollectionMarshal into an ImmutableArray.

Reply:
I had thought using an ArrayBuilder would be conceptually easier, but if you've made it through this PR, you're totally fine with the concept of ImmutableCollectionsMarshal